### ISSUE TITLE
Entwine build memory leak

### ISSUE THREAD

**Issue by mccarthyryanc** [connormanning/entwine#133]
I'm running scan/build steps for ~3k LAZ files using docker:
```bash
docker run --env AWS_ALLOW_INSTANCE_PROFILE=1 \
    --env AWS_DEFAULT_REGION=us-west-2 
    -v ~/:/entwine -it connormanning/entwine \
    scan -t 16 -i s3://bucketname/prefix/ -o /entwine/scan_results.json
```

I modified the `scan_results.json` with the following parameters:
```json
{
  "arbiter":{"s3":{"sse":true}},
  "output":"s3://bucketname/newprefix/",
  "threads": [2,14],
  "resetFiles":10
}
```

And then build:
```
docker run --env AWS_ALLOW_INSTANCE_PROFILE=1 \
    --env AWS_DEFAULT_REGION=us-west-2 \
    -v ~/:/entwine  -it connormanning/entwine \
    build -c /entwine/scan_results_modified.json
```

The `build` process silently fails after adding the 85 file.

I couldn't get a debug build of entwine to test with something like valgrind (#132) but I did track memory usage:
![build_mem_profile](https://user-images.githubusercontent.com/997120/46050361-989a6100-c0e8-11e8-9011-f5211be40812.png)

**Comment by adamsteer** [comment]
commenting to follow - also, this weekend I built a single file with ~650mil points (~5gb as LAZ; ~38gb as LAS). Entwine with -t 1 has so far consumed > 60 GB of RAM (46 physical, 17 swap) to build the index:

`docker pull connormanning/entwine:latest`
`docker run -it -v $(pwd):/opt/data connormanning/entwine  build -i 5gb.laz -o ./output -t 1 —ticks 128`

(-t 1 actually starts 4 processes, my invocation was pretty crude)

This *seems* excessive, I presume entwine has to unpack all the dimensions - and therefore consume roughly the uncompressed file size in memory, plus some overhead - I wasn’t expecting almost double.

Aside from looking for leaks, would creating some rough rules of thumb around expected memory consumption be useful?

**Comment by connormanning** [comment]
@mccarthyryanc thanks for this analysis.  To some extent, this could potentially be the expected behavior.  In the master branch, there exists memory-pooling that grows to the maximal size necessary during the index.  Eventually it would be expected to max out at some point, and without your specific data and timings it's difficult to tell whether your exact memory level is expected or not.  This behavior of Entwine has been altered in the development branch (see below).

@adamsteer some rules of thumb might be useful, and previous rules of thumb I've mentioned are "expect approximately the same number of GB of RAM usage as threads used", but depending on the data this may be inaccurate by a factor of at least 2 in either direction, historically.

FYI - the current development branch is `format` (and corresponding docker tag `connormanning/entwine:format`), which will produce data that is *not* currently readable by potree.entwine.io and speck.ly.  See #134.  Among these formatting changes there have also been a number of (hopefully) optimizations, including removal of the memory-pooling mentioned above.  So any performance analysis would be best served against this branch since this will be merged to master once clients mentioned in #134 are updated.  It should be more performant in general, but if there were already memory issues it's certainly possible they've persisted, so similar analysis would be useful against that branch.  In the past I've not seen any issues from Valgrind analysis on `master`.

I'll be testing this branch independently on much of our currently publicized data over the coming weeks, but third-party analyses are welcome.  Any PRs shown to address issues like this would be welcome.

**Comment by mccarthyryanc** [comment]
@connormanning thanks for the explanation, I didn't quite understand indexing/memory-pooling. It doesn't sound like this is a memory-leak, so I'll close this issue.

**Comment by mccarthyryanc** [comment]
@connormanning and @adamsteer , I ran another `build` test against the format branch but kept the number of threads low and added a subset range:
```bash
docker run --env AWS_ALLOW_INSTANCE_PROFILE=1 \
    --env AWS_DEFAULT_REGION=us-west-2 \
    -v /home/ubuntu/shared:/entwine \
    connormanning/entwine:format build \
    -i /entwine/scan/ept-scan.json -o s3://bucketname/ept \
    --resetFiles 10 \
    --threads 4 \
    --subset 1264 3320
```
![image](https://user-images.githubusercontent.com/997120/46758473-bee00500-cc81-11e8-87ac-b29e888ca549.png)

You'll notice that the subset range isn't a power of 2 or a perfect square, so it looks like it keeps running over files outside the given subset (should have only processed a few files). Aside from that I think the memory usage matches the rule of thumb: number of threads is number of GB. In this case ~4GB, but definitely stayed below the 8GB, even with the subset error. I ended up killing the process, I'm not sure if it would have stopped given that subset range
